src/cockpit/channels/filesystem.py

```diff
@@ -41,53 +48,189 @@ def tag_from_path(path):
         return None
 
 
-def tag_from_fd(fd):
-    try:
-        return tag_from_stat(os.fstat(fd))
-    except OSError:
-        return None
+class AttributeReporter:
+    uid_cache: Dict[int, Union[str, int]]
+    gid_cache: Dict[int, Union[str, int]]
+    report_target: bool
+    getters: Dict[str, Callable[[os.stat_result], JsonDocument]]
 
+    def do_get_tag(self, buf: os.stat_result) -> str:
+        return tag_from_stat(buf)
 
-class FsListChannel(Channel):
-    payload = 'fslist1'
+    def do_get_type(self, buf: os.stat_result) -> str:
+        if stat.S_ISDIR(buf.st_mode):
+            return 'directory'
+        elif stat.S_ISREG(buf.st_mode):
+            return 'file'
+        elif stat.S_ISLNK(buf.st_mode):
+            return 'link'
+        else:
+            return 'special'
+
+    def do_get_mode(self, buf: os.stat_result) -> int:
+        return buf.st_mode & 0o777
+
+    def do_get_owner(self, buf: os.stat_result) -> Union[int, str]:
+        if buf.st_uid not in self.uid_cache:
+            try:
+                self.uid_cache[buf.st_uid] = pwd.getpwuid(buf.st_uid).pw_name
+            except KeyError:
+                self.uid_cache[buf.st_uid] = buf.st_uid
+        return self.uid_cache[buf.st_uid]
+
+    def do_get_group(self, buf: os.stat_result) -> Union[int, str]:
+        if buf.st_gid not in self.gid_cache:
+            try:
+                self.gid_cache[buf.st_gid] = grp.getgrgid(buf.st_gid).gr_name
+            except KeyError:
+                self.gid_cache[buf.st_gid] = buf.st_gid
+        return self.gid_cache[buf.st_gid]
+
+    def do_get_size(self, buf: os.stat_result) -> int:
+        return buf.st_size
+
+    def do_get_modified(self, buf: os.stat_result) -> float:
+        return buf.st_mtime
+
+    def report(self, path: str, dir_fd: int = -1) -> JsonObject:
+        buf = os.lstat(path, dir_fd=dir_fd)
+        attrs = {attr: getter(buf) for attr, getter in self.getters.items()}
+        if self.report_target and stat.S_ISLNK(buf.st_mode):
+            try:
+                attrs['target'] = os.readlink(path, dir_fd=dir_fd)
+            except OSError:
+                pass  # no need to impede reporting the other attributes...
+        return attrs
+
+    def __init__(self, attrs: Sequence[str]) -> None:
+        self.uid_cache = {}
+        self.gid_cache = {}
+        # 'target' gets handled specially (since it's not in the stat data)
+        self.report_target = 'target' in attrs
+        self.getters = {attr: getattr(self, f'do_get_{attr}') for attr in attrs if attr != 'target'}
```

Comment:
This can easily lead to `KeyError` crashes if the caller specifies an unknown attribute. Or is `attrs` pre-validated here?

Reply:
It should be `AttributeError` and I thought I had a catch for that somewhere...